### ISSUE TITLE
How to use mapped+remote mode with dynamic JSON

### ISSUE THREAD

**Issue by pszemus** [kaltura/nginx-vod-module#1097]
Hi
I want to concatenate multiple mp4 files in one MPEG-DASH stream.
1. Is it possible to configure this module in a way that when user requests for `/dynamic.json/manifest.mpd` nginx will issue a subrequest to `http://adserver/dynamic.json` that returns different response for different users?
2. Is it possible to have clip name in clip's chunks names (not only `fragmentX`) so that if the same clip is in multiple different JSON reponses it's chunks can be cached and served from standard nginx cache?

**Comment by pszemus** [comment]
The idea that I have in mind right now:
1. player asks nginx-vod-module (mapped mode) for MPEG-DASH `/manifest.mpd`
   1. nginx-vod-module asks `http://adserver/dynamic.json` for ads for that user
   1. nginx-vod-module receives `dynamic.json` and generates `manifest.mpd`
1. player asks nginx-vod-module (remote mode) for MPEG-DASH chunks: `/media/ad1.mp4/fragment-1-v1-x3.m4s`
   1. nginx-vod-module asks `http://mp4storage/ad1.mp4` for `/media/ad1.mp4/fragment-1-v1-x3.m4s`

In order it to work:
- both locations (`/` and `/media`) have to have access to the same mp4 files
- `manifest.mpd` returned by location `/` (mapped mode) has to have:
   - `media` and `initialization` paths prefixed with `/media/$mp4filename`
   - chunk numbering starting from 1

That way the `dynamic.json` is requested just once, I could cache all the chunks, and they are immutable regardless of their corresponding mp4 file's position in `dynamic.json`.

What do you think about that? Chunk numbering and paths could be done in lua module if nginx-vod-module does not have necessary configuration options.

**Comment by erankor** [comment]
The module outputs continuous timestamps, even in multi-period DASH, so the segments will not be same if they are part of different MPDs. 
Sounds to me that it may be easiest to have the module perform the segmentation of the media files, and perform the manifest manipulation elsewhere (maybe in some high level language like node.js)

**Comment by pszemus** [comment]
I prepared a proof-of-concept and it works in Google Shaka Player and MPEG-DASH reference player.

1. I'm downloading a multi-period manifest from:
```
		location / {
			vod dash;
			vod_mode mapped;

			# location of JSON
			vod_upstream_location /playlist;

			# location of source video files
			vod_remote_upstream_location /media-origin;

			# TODO:
			# - @media and @initialization attributes should be prefixed with /media/$mp4filename/
			# - @startNumber attribute should be set to 1
			vod_base_url "";
		}
```
2. Then I alter the manifest.mpd and change each `<Period>`'s `@startNumber` to 1 and set base URL for `@initialization` and `@media` attributes to `/media/<put mp4 file name here>`
3. Segments are downloaded from `/media/`:
```
		location /media/ {
			vod dash;
			vod_mode remote;
			vod_upstream_location /media-origin;
			vod_base_url ""; # relative segment URLs
			rewrite ^/media(.*)$ $1 break;
		}
```

As I said before, this stream can be played in Google Shaka Player and MPEG-DASH reference player.

I checked segment file structure and compared to other segments:
```
diff <(mp4dump fragment-1-v1-x3.m4s) <(mp4dump fragment-4-v1-x3.m4s)
9c9
<     sequence number = 0
---
>     sequence number = 3
18c18
< [mdat] size=8+1155459
---
> [mdat] size=8+968849
```
I reckon that there are no timestamp differences between segments, only sequence number changes.

To sum up, in order to accomplish dynamic JSON and static segments, nginx-vod-module should have:
- the possibility to start segment numbering from 1 for every `<Period>` :
   - set `@startnumber` to 1 
   - segment's `moof/mfhd/sequence number` should start from 1
- the possibility to set `vod_base_url` with mp4 file name so that initialization and media URL be different for every `<Period>`, eg:
```
media="/media/rek-g2.mp4/fragment-$Number$-$RepresentationID$.m4s"
initialization="/media/rek-g2.mp4/init-$RepresentationID$.mp4"
```
